lib/net/FetchHttpClient.js

```diff
@@ -0,0 +1,122 @@
+  }
+
+  /* eslint-disable class-methods-use-this */
+  /** @override. */
+  getClientName() {
+    return 'fetch';
+  }
+  /* eslint-enable class-methods-use-this */
+
+  makeRequest(
+    host,
+    port,
+    path,
+    method,
+    headers,
+    requestData,
+    protocol,
+    timeout
+  ) {
+    const isInsecureConnection = protocol === 'http';
+
+    const url = new URL(
+      path,
+      `${isInsecureConnection ? 'http' : 'https'}://${host}`
+    );
+    url.port = port;
+
+    const fetchPromise = this._fetchFn(url.toString(), {
+      method,
+      headers,
+      body: requestData || undefined,
+    });
+
+    // The Fetch API does not support passing in a timeout natively, so a
+    // timeout promise is constructed to race against the fetch and preempt the
+    // request, simulating a timeout.
+    //
+    // This timeout behavior differs from Node:
+    // - Fetch uses a single timeout for the entire length of the request.
+    // - Node is more fine-grained and resets the timeout after each stage of
+    //   the request.
+    //
+    // As an example, if the timeout is set to 30s and the connection takes 20s
+    // to be established followed by 20s for the body, Fetch would timeout but
+    // Node would not. The more fine-grained timeout cannot be implemented with
+    // fetch.
+    let pendingTimeoutId;
+    const timeoutPromise = new Promise((_, reject) => {
+      pendingTimeoutId = setTimeout(() => {
+        pendingTimeoutId = null;
+        reject(HttpClient.makeTimeoutError());
+      }, timeout);
+    });
+
+    return Promise.race([fetchPromise, timeoutPromise])
+      .then((res) => {
+        return new FetchHttpClientResponse(res);
+      })
+      .finally(() => {
+        if (pendingTimeoutId) {
+          clearTimeout(pendingTimeoutId);
+        }
+      });
+  }
+}
+
+class FetchHttpClientResponse extends HttpClientResponse {
+  constructor(res) {
+    super(
+      res.status,
+      FetchHttpClientResponse._transformHeadersToObject(res.headers)
+    );
+    this._res = res;
+  }
+
+  getRawResponse() {
+    return this._res;
+  }
+
+  toStream(streamCompleteCallback) {
+    // Unfortunately `fetch` does not have event handlers for when the stream is
+    // completely read. We therefore invoke the streamCompleteCallback right
+    // away. This callback emits a response event with metadata and completes
+    // metrics, so it's ok to do this without waiting for the stream to be
+    // completely read.
+    streamCompleteCallback();
+
+    // Fetch's `body` property is expected to be a readable stream of the body.
+    return this._res.body;
+  }
+
+  toJSON() {
+    return this._res.json();
+  }
+
+  static _transformHeadersToObject(headers) {
+    // Fetch uses a Headers instance so this must be converted to a barebones
+    // JS object to meet the HttpClient interface.
+    const headersObj = {};
+
+    for (const entry of headers) {
+      headersObj[entry[0]] = entry[1];
```

Comment:
WDYT of throwing an exception here if `entry` is not an array with two entries? 
I think maybe it's worth being more defensive here because `fetchFn` is injected

Reply:
Good call! Done.